jerry-core/ecma/operations/ecma-big-uint.c

```diff
@@ -1410,71 +1488,91 @@ ecma_big_uint_shift_right (ecma_extended_primitive_t *left_value_p,  /**< left B
   }
   while (result_p > end_p);
 
-  return result_value_p;
+  if (JERRY_LIKELY (!increase_result))
+  {
+    return result_value_p;
+  }
+  return ecma_big_uint_increase_result (result_value_p);
 } /* ecma_big_uint_shift_right */
 
+#if ENABLED (JERRY_ESNEXT)
+
 /**
- * Helper function for bitwise operations which drops leading zeroes
+ * Compute the left value raised to the power of right value
  *
- * @return ecma BigInt value or ECMA_VALUE_ERROR
- *         Returned value must be freed with ecma_free_value.
+ * return new BigUInt value, NULL on error
  */
-static ecma_extended_primitive_t *
-ecma_big_uint_normalize_result (ecma_extended_primitive_t *value_p, /**< BigUInt value */
-                                ecma_bigint_digit_t *last_digit_p) /**< points to the end of BigUInt */
+ecma_extended_primitive_t *
+ecma_big_uint_pow (ecma_extended_primitive_t *left_value_p, /**< left BigUInt value */
+                   uint32_t right_value) /**< power value */
 {
-  JERRY_ASSERT (last_digit_p[-1] == 0);
+  ecma_extended_primitive_t *result_p = (right_value & 0x1) ? left_value_p : NULL;
+  ecma_extended_primitive_t *square_p = left_value_p;
 
-  ecma_bigint_digit_t *first_digit_p = ECMA_BIGINT_GET_DIGITS (value_p, 0);
+  JERRY_ASSERT (right_value >= 2);
 
-  do
+  while (true)
   {
-    --last_digit_p;
-  }
-  while (last_digit_p > first_digit_p && last_digit_p[-1] == 0);
+    ecma_extended_primitive_t *new_square_p = ecma_big_uint_mul (square_p, square_p);
 
-  return ecma_big_uint_shrink_value (value_p, (uint32_t) ((uint8_t *) last_digit_p - (uint8_t *) first_digit_p));
-} /* ecma_big_uint_normalize_result */
+    if (JERRY_UNLIKELY (new_square_p == NULL))
+    {
+      if (result_p != NULL && result_p != left_value_p)
+      {
+        ecma_deref_bigint (result_p);
+      }
+      result_p = NULL;
+      break;
+    }
 
-/**
- * Helper function for bitwise operations which increases the result by 1
- *
- * @return new BigInt value, NULL on error
- */
-static ecma_extended_primitive_t *
-ecma_big_uint_increase_result (ecma_extended_primitive_t *value_p) /**< BigUInt value */
-{
-  uint32_t size = ECMA_BIGINT_GET_SIZE (value_p);
+    if (square_p != left_value_p)
+    {
+      ecma_deref_bigint (square_p);
+    }
 
-  JERRY_ASSERT (size > 0);
+    square_p = new_square_p;
+    right_value >>= 1;
 
-  ecma_bigint_digit_t *first_digit_p = ECMA_BIGINT_GET_DIGITS (value_p, 0);
-  ecma_bigint_digit_t *last_digit_p = ECMA_BIGINT_GET_DIGITS (value_p, size);
+    if ((right_value & 0x1) != 0)
```

Comment:
Magic number literal, `(right_value & ECMA_BIGINT_SIGN)`

Reply:
This is also not related to sign, it is checked whether the last bit is 1. I can add a macro.